From 2e3f1f0991163004ea441307ab52ce4ea3e068d7 Mon Sep 17 00:00:00 2001
From: Eric Anholt <eric@anholt.net>
Date: Mon, 14 Jan 2019 12:35:43 -0800
Subject: [PATCH 587/703] drm/v3d: Skip MMU flush if the device is currently
 off.

If it's off, we know it will be reset on poweron, so the MMU won't
have any TLB cached from before this point.  Avoids failed waits for
MMU flush to reply.

Signed-off-by: Eric Anholt <eric@anholt.net>
(cherry picked from commit 3ee4e2e0a9e9587eacbb69b067bbc72ab2cdc47b)
---
 drivers/gpu/drm/v3d/v3d_mmu.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/drivers/gpu/drm/v3d/v3d_mmu.c
+++ b/drivers/gpu/drm/v3d/v3d_mmu.c
@@ -18,6 +18,8 @@
  * each client.  This is not yet implemented.
  */
 
+#include <linux/pm_runtime.h>
+
 #include "v3d_drv.h"
 #include "v3d_regs.h"
 
@@ -34,6 +36,14 @@ static int v3d_mmu_flush_all(struct v3d_
 {
 	int ret;
 
+	/* Keep power on the device on until we're done with this
+	 * call, but skip the flush if the device is off and will be
+	 * reset when powered back on.
+	 */
+	ret = pm_runtime_get_if_in_use(v3d->dev);
+	if (ret == 0)
+		return 0;
+
 	/* Make sure that another flush isn't already running when we
 	 * start this one.
 	 */
@@ -61,6 +71,9 @@ static int v3d_mmu_flush_all(struct v3d_
 	if (ret)
 		dev_err(v3d->dev, "MMUC flush wait idle failed\n");
 
+	pm_runtime_mark_last_busy(v3d->dev);
+	pm_runtime_put_autosuspend(v3d->dev);
+
 	return ret;
 }
 
